### PR TITLE
Feat/init main entrypoint proof mode

### LINKED ISSUE #342: feat: complete `initMainEntrypoint` for `CairoRunner`
### Issue Description

Currently, our `initMainEntrypoint` function within the `CairoRunner` module is incomplete and lacks proper handling for the proof mode. This is the source file in question: 

https://github.com/keep-starknet-strange/ziggy-starkdust/blob/698b77942879127dd1e0595ca750ba694073b5fb/src/vm/runners/cairo_runner.zig#L232-L252

Depends on #343.

### Action Items

1. **Completion of Implementation:**
   - Refer to the Rust implementation for guidance: https://github.com/lambdaclass/cairo-vm/blob/4ea224a1068b43aaa3899d24e834475106e8a78f/vm/src/vm/runners/cairo_runner.rs#L471-L561.
   - Ensure the function is implemented correctly, especially with regard to handling the proof mode.

2. **Unit Test Suite:**
   - Develop a comprehensive suite of unit tests to validate the proper implementation of the `initMainEntrypoint` function.
   - Refer to the Rust reference for tests: https://github.com/lambdaclass/cairo-vm/blob/4ea224a1068b43aaa3899d24e834475106e8a78f/vm/src/vm/runners/cairo_runner.rs#L1832-L1851.

### LINKED ISSUE #343: feat: add `isProofMode` utility function to `CairoRunner`
### Issue Description

We require the addition of a utility function, `isProofMode`, within the `CairoRunner` struct. This utility function will serve to check whether the application is currently running in proof mode. 

### Action Items

**Implementation:**
   - Refer to the Rust implementation for guidance: https://github.com/lambdaclass/cairo-vm/blob/4ea224a1068b43aaa3899d24e834475106e8a78f/vm/src/vm/runners/cairo_runner.rs#L318-L321.
   - Ensure the implementation aligns with the existing logic in the Rust reference.

### PR DESCRIPTION
**Pull request by StringNick**
closes #342 
closes #343

### CODE REVIEW

File: src/vm/runners/cairo_runner.zig [hunk 1]
```diff
@@ -216,39 +222,81 @@ pub const CairoRunner = struct {
...
+                try stack_prefix.append(MaybeRelocatable.fromRelocatable(try self.execution_base.addUint(target_offset)));
+                try stack_prefix.appendSlice(stack.items);
+
+                var execution_public_memory = try std.ArrayList(usize).initCapacity(self.allocator, stack_prefix.items.len);
+                for (0..stack_prefix.items.len) |v| {
+                    try execution_public_memory.append(v);
+                }
+                self.execution_public_memory = execution_public_memory;
+
+                try self.initState(try (if (self.program.getStartPc()) |start| start else RunnerError.NoProgramStart), &stack_prefix);
```
Comment (tcoratger) [inline src/vm/runners/cairo_runner.zig:266]: Should be 

```zig
try self.initState(self.program.getStartPc() orelse return RunnerError.NoProgramStart, &stack_prefix);
```

no?

File: src/vm/runners/cairo_runner.zig [hunk 2]
```diff
@@ -216,39 +222,81 @@ pub const CairoRunner = struct {
...
+            } else {
+                target_offset = stack.items.len + 2;
+
+                const return_fp = try self.vm.segments.addSegment();
+                const end = try self.vm.segments.addSegment();
+                try stack.append(MaybeRelocatable.fromRelocatable(return_fp));
+                try stack.append(MaybeRelocatable.fromRelocatable(end));
+
+                try self.initState(try if (self.program.getStartPc()) |start| start else RunnerError.NoProgramStart, &stack);
```
Comment (tcoratger) [inline src/vm/runners/cairo_runner.zig:275]: Same here

File: src/vm/runners/cairo_runner.zig [hunk 3]
```diff
@@ -216,39 +222,81 @@ pub const CairoRunner = struct {
...
+            }
+
+            self.initial_fp = try self.execution_base.addUint(target_offset);
+            self.initial_ap = self.initial_fp;
+
+            return self.program_base.addUint(try if (self.program.getEndPc()) |end| end else RunnerError.NoProgramEnd);
```
Comment (tcoratger) [inline src/vm/runners/cairo_runner.zig:281]: Same here

File: src/vm/runners/cairo_runner.zig [hunk 4]
```diff
@@ -147,17 +150,20 @@ pub const CairoRunner = struct {
             },
             .instructions = instructions,
             .vm = vm,
-            .function_call_stack = std.ArrayList(MaybeRelocatable).init(allocator),
-            .proof_mode = proof_mode,
+            .runner_mode = if (proof_mode) RunnerMode.proof_mode_canonical else RunnerMode.execution_mode,
```
Comment (tcoratger) [inline src/vm/runners/cairo_runner.zig:153]:
```suggestion
            .runner_mode = if (proof_mode) .proof_mode_canonical else .execution_mode,
```

File: src/vm/runners/cairo_runner.zig (hunk 1 above)
Comment (StringNick) [reply src/vm/runners/cairo_runner.zig:266]: yes good solution i forgot about orelse, but circle brackets we need anyway :(

**Comment by StringNick** [comment]
@tcoratger fixed, thank you

**Review by lana-shanghai — APPROVED** [review]
lgtm

**Pull request merged**